File: General-DSA-Questions/Linked-List/6_removeDuplicatesFromSortedLinkedList.cpp
// Write a function to remove duplicates from a sorted linked list. The function should modify the list in place and return the head of the modified list.

#include <iostream>

using namespace std;

struct Node {
    int data;
    Node* next;
    
    Node(int val) : data(val), next(nullptr) {}
};

// Function to remove duplicates from a sorted linked list
Node* removeDuplicates(Node* head) {
    // If the list is empty or has only one node, there are no duplicates
    if (head == nullptr || head->next == nullptr) return head;

    Node* current = head;

    // Traverse the list
    while (current != nullptr && current->next != nullptr) {
        // If the current node's data is the same as the next node's data
        if (current->data == current->next->data) {
            // Skip the next node
            Node* temp = current->next;
            current->next = current->next->next;
            delete temp;
        } else {
            // Move to the next node
            current = current->next;
        }
    }

    return head;
}

// Helper function to print the linked list
void printList(Node* head) {
    Node* current = head;

    while (current != nullptr) {
        cout << current->data << " ";
        current = current->next;
    }

    cout << endl;
}

int main() {
    Node* head = new Node(1);
    head->next = new Node(2);
    head->next->next = new Node(2);
    head->next->next->next = new Node(2);
    head->next->next->next->next = new Node(3);

    printList(head);
    printList(removeDuplicates(head));
}

// Assume that we have given that linked list is not sorted then we have 2 solutions to solve the same questions to duplicate the elements in the linked list.
// 1. Using Hashing(map) - Time complexity: O(N), Space complexity: O(N) - store the elements in the map and check if the element is already present in the map then delete the node.
//    a. Create a map to store the elements of the linked list. like map<int, bool> m;
//    b. Traverse the linked list and check if the element is already present in the map then delete the node.
//    c. If the element is not present in the map then insert the element in the map.
//    d. Return the head of the linked list.
// 2. Using 2 pointers - Time complexity: O(N^2), Space complexity: O(1) - This is the brute force approach

File: JS-DSA-Questions/Linked-List/6_removeDuplicatesFromSortedLinkedList.js
// Remove duplicates from a SORTED linked list, modifying it in place; return the head.

// Intuition:
// 1. The list is sorted, so duplicates are adjacent.
// 2. If current.data === current.next.data, skip the next node.
// 3. Otherwise move to the next node. Continue to the end.

class Node {
  constructor(val) {
    this.data = val;
    this.next = null;
  }
}

function removeDuplicates(head) {
  if (head === null || head.next === null) return head;

  let current = head;

  while (current !== null && current.next !== null) {
    if (current.data === current.next.data) {
      // Skip the duplicate node (JS is garbage-collected, so there is no manual `delete`)
      current.next = current.next.next;
    } else {
      current = current.next;
    }
  }

  return head;
}

function printList(head) {
  const values = [];
  let current = head;
  while (current !== null) {
    values.push(current.data);
    current = current.next;
  }
  console.log(values.join(" "));
}

// Build 1 -> 2 -> 2 -> 2 -> 3
const head = new Node(1);
head.next = new Node(2);
head.next.next = new Node(2);
head.next.next.next = new Node(2);
head.next.next.next.next = new Node(3);

printList(head); // 1 2 2 2 3
printList(removeDuplicates(head)); // 1 2 3

// If the list were NOT sorted, two options:
// 1. Hashing (Set): O(N) time, O(N) space — track seen values and unlink repeats.
// 2. Two pointers: O(N^2) time, O(1) space — brute force.
